### PR TITLE
feat: adds events to handle changes in xblocks

### PR DESCRIPTION
**Pull request by navinkarkera**
**Description:** Adds events and data classes to allow consumers to react on xblock events like published, deleted and duplicated. 
The idea is to fire these events in [delete_item](https://github.com/openedx/edx-platform/blob/4c0b0f99a0d874c589c4dbf1ea44cdfc84edf7f9/lms/djangoapps/ccx/modulestore.py#L237), [publish_item](https://github.com/openedx/edx-platform/blob/4c0b0f99a0d874c589c4dbf1ea44cdfc84edf7f9/lms/djangoapps/ccx/modulestore.py#L263) and [_duplicate_item](https://github.com/openedx/edx-platform/blob/5bc00b373c879fb2859b44d338f81647210c0909/cms/djangoapps/contentstore/views/item.py#L878) and update skills related to the xblock in [taxonomy-connector](https://github.com/openedx/taxonomy-connector) (part of course-discovery).

**Answer for a question asked in previous similar MRs:**

_Do consumers have to make separate requests to get more information?_
**Generally:** Additional data if required related to the event needs to be fetch separately as it can be different for different use case.
**Specifically for skill tagging:** Yes, for skill tagging to work we need concatenated text representation of children xblocks for units and transcript for video xblocks which can be extracted using [index_dictionary](https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/edx_search.html#which-data-gets-indexed) of the xblock. The idea is to add a new api to fetch block metadata and use this API in taxonomy-connector.

**Related MRs**:
These MRs are the foundation for storing skills related on xblocks level, specifically for unit and video xblocks.
- https://github.com/openedx/taxonomy-connector/pull/119
- https://github.com/open-craft/taxonomy-connector/pull/1
- https://github.com/openedx/edx-platform/pull/31273

**ISSUE:** Link to GitHub issue

**Merge checklist:**
- [x] All reviewers approved
- [ ] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** List any concerns about this PR - inelegant
solutions, hacks, quick-and-dirty implementations, concerns about
migrations, etc.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @navinkarkera!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

File: openedx_events/content_authoring/data.py
Comment (tecoholic) [inline openedx_events/content_authoring/data.py:75]: Do we really need 2 data classes here? Can't we simply have just `XBlockData` and have the `source_usage_key` set to `None` for the "non-duplicate" events?

File: openedx_events/content_authoring/signals.py
Comment (tecoholic) [inline openedx_events/content_authoring/signals.py:28]: Nit: for all 3 events
```suggestion
# .. event_description: Fired when an XBlock is published in Studio.
```

File: openedx_events/event_bus/avro/custom_serializers.py
Comment (tecoholic) [inline openedx_events/event_bus/avro/custom_serializers.py:76]:
```suggestion
    CustomTypeAvroSerializer for UsageKey class.
```

**Review by tecoholic — CHANGES_REQUESTED** [review]
@navinkarkera Nice work covering all the necessary parts including the avro tests. I have noted one question about the data classes introduced here and some minor typos. Can you kindly address those?

**Comment by tecoholic** [comment]
@navinkarkera One thing that kind of stood out from the PR description was this question:

> Do consumers have to make separate requests to get more information?

The answer is specific to skill tagging which is great. It can also be more generally answered, as the event is only passing the `UsageKey` to the consumer, all other metadata will need to be fetched by the consumer separately. Eg., User deleting an XBlock, diff of the content during publishing..etc.,

This is just a note, to help myself understand the scope of the change here and aid other reviewers, if needed. No action item here :)

File: openedx_events/content_authoring/data.py
Comment (navinkarkera) [reply openedx_events/content_authoring/data.py:75]: @tecoholic True, just kept it separate as having `source_usage_key` for publish and delete events did not make any sense. Also for duplicate event `source_usage_key` is required which we cannot enforce if it has a default value of `None`.

**Comment by navinkarkera** [comment]
@tecoholic Thanks for the suggestions! Updated the MR.

>The answer is specific to skill tagging which is great. It can also be more generally answered, as the event is only passing the UsageKey to the consumer, all other metadata will need to be fetched by the consumer separately. Eg., User deleting an XBlock, diff of the content during publishing..etc.,

Makes absolute sense. Thanks, i'll add this to the description.

**Comment by navinkarkera** [comment]
@mariajgrimaldi Sorry, when I clicked on request-a-review-again button for @tecoholic, it removed you from reviewers list and I cannot add you back.
>Hello! Thanks for your contribution! I'll review this as soon as I can.

Thank you!

File: openedx_events/content_authoring/data.py
Comment (tecoholic) [reply openedx_events/content_authoring/data.py:75]: @navinkarkera I see. Thank you for the explanation.

**Review by tecoholic — APPROVED** [review]
@navinkarkera :+1:

- [x] I tested this: As this only introduces the data classes and the signals, there wasn't anything specific to test. The unit tests are passing, that validates the data classes.
- [x] I read through the code
- [ ] I checked for accessibility issues - NA
- [x] Includes documentation - It contains inline documentation as expected.

File: openedx_events/content_authoring/data.py
Comment (mariajgrimaldi) [reply openedx_events/content_authoring/data.py:75]: I'm not sure I understand the purpose of XBlockDuplicatedData.

File: openedx_events/content_authoring/data.py
Comment (mariajgrimaldi) [inline openedx_events/content_authoring/data.py:69]: I don't think this comment is relevant here

File: openedx_events/content_authoring/data.py
Comment (mariajgrimaldi) [inline openedx_events/content_authoring/data.py:60]: I think I'd call this just `XBlockData` or do you think there's a difference?

File: openedx_events/event_bus/avro/tests/test_deserializer.py
Comment (mariajgrimaldi) [inline openedx_events/event_bus/avro/tests/test_deserializer.py:105]: The missing docstring is a bit concerning, but I understand you're following the pattern in this test file. So that gives me a bit of work for the future.

**Review by mariajgrimaldi — CHANGES_REQUESTED** [review]

File: openedx_events/content_authoring/data.py
Comment (navinkarkera) [reply openedx_events/content_authoring/data.py:75]: @mariajgrimaldi We want to react in course-discovery/taxonomy whenever an xblock is duplicated.  The idea is to store tags for a xblock in taxonomy-connector (tags are fetched from an external API by sending text representation of the xblock). Now if an xblock is duplicated, we plan to copy the tags over to the new xblock and skip sending to external API. Hence we need the source xblocks usage_key to get the tags from.

File: openedx_events/event_bus/avro/tests/test_deserializer.py
Comment (navinkarkera) [reply openedx_events/event_bus/avro/tests/test_deserializer.py:105]: @mariajgrimaldi Started by adding docstrings in the tests that I created. :)

File: openedx_events/content_authoring/data.py
Comment (mariajgrimaldi) [reply openedx_events/content_authoring/data.py]: Thanks for the explanation! Now, I wonder if we should use DuplicatedXblockData instead of XblockDuplicatedData, since it seems that the 1st indicates data from a duplicated block and the 2nd that the data is duplicated? what do you think?

**Comment by mariajgrimaldi** [comment]
do you have a working example for testing? if so, can you add instructions to the PR description? thanks!

File: openedx_events/content_authoring/data.py
Comment (navinkarkera) [reply openedx_events/content_authoring/data.py]: @mariajgrimaldi Thanks! the name `DuplicatedXBlockData` makes more sense. Updated!

**Comment by navinkarkera** [comment]
> do you have a working example for testing? if so, can you add instructions to the PR description? thanks!

@mariajgrimaldi I don't have one yet as the work on the edx-platform side is pending. I do want to add one more signal and data class to send list of verified skill ids and xblock usage_key. The idea is that users will verify the tags/skills associated to an XBlock. We want to send this data via openedx-event signals to course-discovery and update the relevant tables. I was wondering whether I should add it in the same MR or create a separate MR once this is merged. Do let me know your thoughts here.

**Comment by mariajgrimaldi** [comment]
@navinkarkera: it's best to create a separate PR for those changes.

@felipemontoya: what do you think of these changes? I'd like to test a working version, for now, this looks quite good!

**Comment by navinkarkera** [comment]
@mariajgrimaldi Thanks! I have created a separate MR: https://github.com/open-craft/openedx-events/pull/1 on top of this branch. The new event requires support for array types so it is also added in the MR. Please have a look.

**Comment by navinkarkera** [comment]
@mariajgrimaldi Any updates on this? Let me know if something needs to be changed.

**Comment by felipemontoya** [comment]
Hello @navinkarkera, thank for the contribution.

Even though I came in late to the discussion, everything seems taken care of.
You even added the avro serializer. The only question I have remaining is if there is any need to also pass the user reference to the events? In the publish, delete and duplicate actions that you linked there is always an xblock and a user. I don't necessarily think that we should pass it along, but I'l like to know if that user is important.

**Comment by navinkarkera** [comment]
> Hello @navinkarkera, thank for the contribution.
> 
> Even though I came in late to the discussion, everything seems taken care of. You even added the avro serializer. The only question I have remaining is if there is any need to also pass the user reference to the events? In the publish, delete and duplicate actions that you linked there is always an xblock and a user. I don't necessarily think that we should pass it along, but I'l like to know if that user is important.

@felipemontoya Thanks. We don't need user details, at least for the purpose that I am adding these events for.

**Review by felipemontoya — APPROVED** [review]
Thanks @navinkarkera, I think the current revision is ready to continue

File: openedx_events/content_authoring/signals.py
Comment (ormsbee) [inline openedx_events/content_authoring/signals.py:28]: Quick note: Please be very specific in the description as to exactly when this is fired. For instance:

1. Does it only fire for XBlocks that are changing in the `published` branch, or will it also fire for changes to XBlocks in the draft branch?
2. Does it fire only for when something is initially created, or also on updates?
3. Does it fire for changes that happen to content in content libraries?
4. Does it fire when a parent XBlock has changed, potentially changing the value of inherited fields?
5. Does it fire when a child XBlock has changed (e.g. when it fires for a problem, does it also fire for the Unit it's enclosed in)?

I honestly don't want to engage in a long discussion about when it _should_ fire. I'm guessing you already have some part of the code that you're plugging into that will address your use cases, and I don't want to complicate your implementation. But please be as precise as you can in both the naming and description of these signals.

Comment (ormsbee) [reply openedx_events/content_authoring/signals.py:28]: (@navinkarkera: Please take this as a non-binding comment. You already have approval from @felipemontoya, who is a maintainer of this repo.)

Comment (ormsbee) [reply openedx_events/content_authoring/signals.py:28]: > If a unit is published, this event will be fired with only its own usage_key, no additional events are published for child or parent blocks.

I see. So if publishing happens at the section or subsection level, there is likewise no event fired for the unit? This would be good to document in the description so people aren't caught by surprise.

Comment (navinkarkera) [reply openedx_events/content_authoring/signals.py:28]: @ormsbee So if a subsection is published, an event is fired with `block_type=sequential` and if a section is published, event with `block_type=chapter` is fired. Should I include something like:

If a parent block with changes in one or more child blocks is published, only a single `XBLOCK_PUBLISHED` event is fired with parent block details. For example: If a section is published with changes in multiple units, only a single event is fired with section details like : `XBlockData(usage_key="section-usage-key", block_type="chapter")`

Comment (ormsbee) [reply openedx_events/content_authoring/signals.py:28]: Yeah, I think that would be a really helpful comment to have in that annotation. Thank you!

**Comment by mariajgrimaldi** [comment]
@navinkarkera: can you fix the CI? I think a test for the new type is missing. Do you know if `UsageKey` is ever used [here](https://github.com/openedx/openedx-events/pull/143/files#diff-39836b2946ae979f15d1ed71bf67abd997524c61b3e366c248d842e5e932e494R57-R64)?

**Comment by navinkarkera** [comment]
> @navinkarkera: can you fix the CI? I think a test for the new type is missing. Do you know if `UsageKey` is ever used [here](https://github.com/openedx/openedx-events/pull/143/files#diff-39836b2946ae979f15d1ed71bf67abd997524c61b3e366c248d842e5e932e494R57-R64)?

@mariajgrimaldi Yes, it is used in https://github.com/openedx/openedx-events/blob/1934a839b6b04db1b02634cd74320045f2e6ed17/openedx_events/event_bus/avro/tests/test_avro.py#L70-L83

Not sure what is missing now.

**Comment by navinkarkera** [comment]
@mariajgrimaldi Gentle reminder.

**Comment by mariajgrimaldi** [comment]
@navinkarkera: I'm out for a few days, I'll check on this when I'm back. Sorry for the delay.

**Comment by mariajgrimaldi** [comment]
can you rebase with main? thank you!

**Comment by navinkarkera** [comment]
> can you rebase with main? thank you!

@mariajgrimaldi It is already up to date.

**Comment by mariajgrimaldi** [comment]
I thought rebasing would help with the coverage issue. Can you check that so we can merge? Let me know if you need help!

**Comment by navinkarkera** [comment]
> I thought rebasing would help with the coverage issue. Can you check that so we can merge? Let me know if you need help!

@mariajgrimaldi Running `make coverage` locally shows that the coverage increased from 95% to 96%. Also visiting the [codecov link](https://app.codecov.io/gh/openedx/openedx-events/pull/143) also shows an increase of 0.13% as opposed to what is displayed here in github below.

image

Not sure how to proceed here. Should we implement tests for missing lines even if they are not related to any change in this MR? Please advice.

**Comment by mphilbrick211** [comment]
@mariajgrimaldi Friendly ping on @navinkarkera's comment above.

**Comment by mariajgrimaldi** [comment]
I'm back on monday. I'll take a look at this then. Sorry for the delay!

**Comment by mariajgrimaldi** [comment]
hello! can you rebase with main? The codecov for the project changed, so I think it'll pass this time

**Comment by navinkarkera** [comment]
@mariajgrimaldi Updated! Please approve the workflow.

**Review by mariajgrimaldi — APPROVED** [review]
This looks great! Just one last thing, can you update the changelog entry date? thanks!

**Comment by navinkarkera** [comment]
@mariajgrimaldi Thanks! Updated.

**Comment by openedx-webhooks** [comment]
@navinkarkera 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-events&entry.867055334=https://github.com/openedx/openedx-events/pull/143&entry.1484655318=https://github.com/navinkarkera&entry.752974735=2022-11-08+12:31&entry.1917517419=2023-01-03+14:54&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by mariajgrimaldi** [comment]
New release: https://github.com/openedx/openedx-events/releases/tag/v4.1.0

**Pull request merged**
